### PR TITLE
Feature: Categorical DataFrame

### PR DESCRIPTION
**Pull request by Magnus167**
Example code:
```python
from macrosynergy.management.types import QuantamentalDataFrame
from macrosynergy.management.simulate import make_test_df
from macrosynergy.management.utils import reduce_df
from typing import List
import pandas as pd

cids = ["USD", "EUR", "JPY", "GBP", "AUD", "CAD", "CHF", "SEK"]
xcats = ["FX", "IR", "EQ", "CR", "CO", "FI", "IN", "CM", "ME", "OT"]


df: pd.DataFrame = make_test_df(cids=cids, xcats=xcats)

dfc = df.copy()
print(f"Size of pd.DataFrame: {df.memory_usage(deep=True).sum() / 1024 ** 2} MB")

df = QuantamentalDataFrame(df)  # memory free!

qdf = QuantamentalDataFrame(dfc)  # new copy created, no memory freed

print(f"Size of QDF: {qdf.memory_usage(deep=True).sum() / 1024 ** 2} MB")
print("Is Categorical:", qdf.is_categorical())

red_df = qdf.reduce_df(cids=["USD", "EUR"], xcats=["FX", "IR"])

red_df.reduce_df(cids=["USD"], xcats=["FX"], inplace=True)

wdf: pd.DataFrame = qdf.to_wide()

try:
    test_if_qdf = QuantamentalDataFrame(wdf)
except Exception as e:
    print(f"Error: {e}")
    print("-- Expected : DataFrame is not a QuantamentalDataFrame --")


test_df = QuantamentalDataFrame(df)


def example_func(test_df: QuantamentalDataFrame, cids: List[str]) -> pd.DataFrame:
    if not isinstance(test_df, QuantamentalDataFrame):
        raise TypeError("Expected: QuantamentalDataFrame")
    
    if type(test_df) is QuantamentalDataFrame:
        return test_df.reduce_df(cids=cids)
    else:
        return reduce_df(df=test_df, cids=cids)

```

### CODE REVIEW

File: macrosynergy/management/types/qdf/methods.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+) -> QuantamentalDataFrameBase:
+    """
+    Method used to update the DataFrame on the category level.
+    """
+    if not isinstance(df, QuantamentalDataFrameBase):
+        raise TypeError("`df` must be a QuantamentalDataFrame.")
+    if not isinstance(df_add, QuantamentalDataFrameBase):
+        raise TypeError("`df_add` must be a QuantamentalDataFrame.")
+
+    incumbent_categories = list(df["xcat"].unique())
```
Comment (sandresen1) [inline macrosynergy/management/types/qdf/methods.py:220]: I think you could just use `set(df["xcat"].unique())` here and then not use the set later on

File: macrosynergy/management/types/qdf/base.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        result: bool = True
+        try:
+            # the try except offers a safety net in case the instance is not a
+            # pd.DataFrame and one of the checks raises an error
+            result = result and isinstance(instance, pd.DataFrame)
+            result = result and instance.index.name is None
+            result = result and not isinstance(instance.columns, pd.MultiIndex)
+            result = result and all([col in instance.columns for col in IDX_COLS])
+            result = result and len(instance.columns) > len(IDX_COLS)
+            result = result and len(instance.columns) == len(set(instance.columns))
```
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/management/types/qdf/base.py:28]: I did think of that, but then it makes debugging a bit hard in terms of pointing out which condition actually failed

File: macrosynergy/management/types/qdf/classes.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        if inplace:
+            self.__init__(result, categorical=self.InitializedAsCategorical)
+            return self
+        return QuantamentalDataFrame(result, categorical=self.InitializedAsCategorical)
+
+    def is_categorical(self) -> bool:
+        """
+        Returns True if the QuantamentalDataFrame is categorical.
+        """
+        strcols = list(set(QuantamentalDataFrame.IndexCols) - {"real_date"})
```
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/management/types/qdf/classes.py:53]: almost - but those only account for cid, xcat - this is for any str cols that one might add (label, group, economic area etc.).
unless we want to leave those unmodified?

File: macrosynergy/management/types/qdf/methods.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        except:
+            warnings.warn(
+                f"Could not convert column {col} to {dtype} from {curr_type}."
+            )
+
+    return df
+
+
+def check_is_categorical(df: QuantamentalDataFrameBase) -> bool:
+    if not isinstance(df, QuantamentalDataFrameBase):
```
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/management/types/qdf/methods.py:36]: done

File: macrosynergy/management/types/qdf/classes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/management/types/qdf/classes.py:53]: changed to use the backend method

File: macrosynergy/management/types/qdf/classes.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        for col in strcols:
+            if col in self.columns:
+                if not self[col].dtype.name == "category":
+                    return False
+
+        return True
+
+    def to_categorical(self) -> "QuantamentalDataFrame":
```
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/management/types/qdf/classes.py:61]: yes, will be implementing soon. it simply works if you do:

`from __future__ import annotations`

shouldn't be an issue in python3.10+

File: macrosynergy/management/types/qdf/classes.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
+"""
```
Comment (emhbrine) [inline macrosynergy/management/types/qdf/classes.py:1]: Let's rename this file `quantamental_dataframe`

**Review by emhbrine — APPROVED** [review]

**Review by emhbrine — APPROVED** [review]

**Review by emhbrine — DISMISSED** [review]

**Pull request merged**